Feat import export binary and decimal seed

nosedam opened the pull request:
## Description

This change allows to export and import the seed phrase in binary and decimal format.
Mentioned here: https://github.com/SeedSigner/seedsigner/issues/68

https://github.com/user-attachments/assets/fbb5ff43-19ab-4228-88d1-4a4ab38b0dbe

https://github.com/user-attachments/assets/05ad84c8-5536-4660-8d6d-4f3e777bafaa

https://github.com/user-attachments/assets/444a31ea-cb51-4ee7-ba48-6cfcebba3667

https://github.com/user-attachments/assets/a6c1a8da-f86b-40df-9753-91b0f4e27dcb


This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [X] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [X] N/A (it uses the same original flows, it just uses new keyboards to translate their values to a mnemonic word and continue with original flow)

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [X] Other (Linux VM using Seedsigner emulator)

Thanks everyone for checking the PR. All feedback is appreciated!
Any changes that I need to do just tell me!

nosedam committed:
feat: add export functionality for binary and decimal

nosedam committed:
refactor: rename seed format transformers

nosedam committed:
feat: add decimal seed import functionality

nosedam committed:
feat: add binary import option

nosedam committed:
fix: review details view on other seeds formats

nosedam committed:
feat: add binary keyboard buttons for aditional hardware buttons

nosedam committed:
refactor: rename variables

nosedam committed:
feat: format

fedebuyito commented:
Hi @nosedam . Great job! I just tested your commit and it does what PR states and issue asks. Further, I tested on 3.10 and 3.12 pytest, and both passed.

nosedam commented:
Thanks for trying it @fedebuyito! I will try to improve the binary numbers export and validation by introducing space separation or different colors, so Its easier to identify which digit you are reading.

nosedam committed:
feat: add colors to binary view export to prevent confusion

nosedam commented:
Hi guys, I added a color feature to prevent confusions in seed binary export
![image](https://github.com/user-attachments/assets/79ad8494-c6cc-490b-9257-66b1490a0af8)

fedebuyito commented:
Hi @nosedam . That seems to be fine! 

![image](https://github.com/user-attachments/assets/327c244d-7b83-4f68-b8a4-e299f17279da)

![image](https://github.com/user-attachments/assets/0a55afc6-7806-48ff-842d-7a66062d9aa0)

Pytested, too. All passed.

![image](https://github.com/user-attachments/assets/0be418a0-7b7b-4539-937a-abea5b890c26)
![image](https://github.com/user-attachments/assets/39b393a8-cc67-4cad-b2c5-9fdbd08fbc2b)

kdmukai commented:
@alvroble : waiting for an update on this PR.

alvroble commented:
@nosedam: I wanted to ask if you plan to continue working on this PR, since there’s interest in getting this functionality merged. If you’re not planning to pick it back up, we’ll probably need to open a new PR, cherry-pick your commits, and resolve the conflicts. Thanks!

nosedam commented:
Hi @alvroble! Awesome, yes sure I can work on solving the conflicts, most probably on the weekend. Thanks for the interest on this!

kdmukai commented:
@nosedam some broader context: As SeedSigner pushes hard to include as many languages as possible, we're going to have to start seriously considering how we solve the problem of backing up seeds. I think the general consensus across the bitcoin builders I've spoken with is that we are just stuck with the English bip39 wordlist.

Yes, wordlists in other languages are in bip39, but with basically zero adoption of them AND zero appetite by the BIP editors to have ongoing maintenance to add new languages, we end up with no central "schelling point" to establish new wordlists.

Yet it's obviously unacceptable to expect non-English speakers to secure their life savings via English mnemonic backups, especially for the languages that don't use western characters and even worse for languages like Arabic that write right-to-left.

So... the index numbers for the English bip39 wordlist seem like the best compromise we have right now. Inherently assume that the index numbers map to the English words, but have non-English speakers write backups in numeric form. Obviously they lose all the benefits of human-friendly mnemonic words ("layer" vs "maze" is WAY easier to discern than `1011` vs `1101`), but so far I haven't seen any better proposals for global backups.

_Aside: SeedQR is arguably vastly superior to writing down index numbers since the QR format includes 7% error correction. But I'm not quite ready to advocate for that as the primary global backup strategy. But if SeedSigner does gain worldwide traction, odds are that lots of people will make a SeedQR anyway._

This is all a long introduction to my next ask for this PR:

I very much want "decimal" import / export now. Clearly from the above I view it as a crucial component of global bitcoin self-custody.

Though I dislike the use of the word "decimal"; it's just too technical / mathy and is not immediately intuitive. I'd prefer the phrase "Index number".

However, I remain less enthusiastic or indifferent about binary import / export. The UX color hinting helps, but that in itself will need a deeper design review / redesign by our UX designer.

I think it will simplify the PR review and merge process if you split the binary functionality out into its own PR. Let that part be debated on its own merits and further refined without slowing down the process of merging in the decimal / index number functionality.

kdmukai commented:
For index number display and entry: it feels like we should insist on having leading zeros.

But I'm open to hear others' opinions on this.

nosedam committed:
Merge branch 'dev' into feat-import-export-binary-and-decimal-seed

nosedam committed:
feat: reset to remote

nosedam committed:
Merge branch 'dev' into feat-import-export-binary-and-decimal-seed

nosedam committed:
fix: button format

nosedam committed:
fix: transformers

nosedam committed:
feat: rename decimal to index numbers

nosedam commented:
Hi! Thanks a lot for providing the context! Yeah I really love the SeedQR functionality its really clever and works awesome. 

As for the seed exports, I fixed the merge conflicts and removed the binary export functionality. I left the "seed format transformers" in case its needed to add new formats more easily (I will add a PR with the binary format so it can be discussed separately)

I also renamed "decimal" and changed it for index numbers so its more clear to end users. 
I also added the leading zeros to always have 4 digits (it looks a lot better now!)

<img width="471" height="279" alt="image" src="https://github.com/user-attachments/assets/f3f7b023-6cb2-4445-831f-d22905ce69d0" />
<img width="480" height="263" alt="image" src="https://github.com/user-attachments/assets/f70f45d6-b307-49d8-a924-10c9cde0f0c2" />
<img width="479" height="258" alt="image" src="https://github.com/user-attachments/assets/6edb3450-8159-4f84-bdb1-772ca0bf8532" />
<img width="474" height="256" alt="image" src="https://github.com/user-attachments/assets/497a2a4f-2dfc-40a3-b95d-4e1bbefc5387" />

Thanks a lot for your comments!
Happy to help in anything else thats needed

kriswilk commented:
In the second screenshot, the button label "View seed index number" is rather long. Furthermore, it should be plural (_numbers_).

Since the page is titled "Backup Seed", the user already knows that the context for the choice is related to the seed. I suggest that the list be simplified to:

- View as words
- View as numbers
- Export as SeedQR

This wording emphasizes that the views are merely different representations of the same seed. If/when the binary option is added, it would be "View as binary". Simple and consistent.

nosedam committed:
feat: rename backup seed options

nosedam commented:
Hi! Yeah thats a good suggestion, the other text was too large and needed to be scrolled to be shown fully. Thanks!
It should now show up like this

<img width="471" height="287" alt="image" src="https://github.com/user-attachments/assets/794c2c0f-941e-48a7-abf1-23829f1e145c" />

dataresilience commented:
> For index number display and entry: it feels like we should insist on having leading zeros.
> 
> But I'm open to hear others' opinions on this.

Also keep in mind that the decimal number serving as the _index_ of the BIP-0039 only makes sense if it starts at 0001 - abandon up to and including 2048 - zoo (there is no index 0000 word indexed in BIP-0039) whereas in a binary format it only makes sense to have use the 0000 to 2047 range meaning 11 bits range from 00000000000 to 11111111111 (2047) as 2048 would mean switching on bit 12 and that is outside the scope of the binary seed. If that makes sense? I am eagerly waiting for the index/decimal 0001-2048 feature/option in the seedsigner UX, it is a great suggestion!!! I was really happy to find this pull request here!!! Thank You!

nosedam commented:
Hi! I have no problem in changing this so it starts on number 1 instead of 0 for seed numbers, if everyone agrees I will upload the change.

dataresilience commented:
@nosedam that is great! Perhaps even call it "seed index number" (too long maybe?) in the UX or something similar to make the point that it refers to the BIP-0039 index number? Have a great day!

newtonick commented:
@nosedam 

Sorry this PR has taken so long to get a proper look at by me.

As it stands, it isn’t mergeable because it includes unrelated emulator code and GIF/JPEG/PNG assets that need to be removed. A few other important items also need addressing:

- “View as numbers” bypasses the private-seed warning shown before viewing seed words.
- The binary-entry code remains but references missing helper functions.
- The “Seed Word #…” title lost its translation wrapper.
- The new index-number import/export flows need focused tests.

Do you still want to pursue this PR? If so, please rebase and clean it up around the index-number functionality. If there’s no response in about a month, we’ll close it for now.

dataresilience commented:
Hello! Was showing seed indexes included or dropped? Sorry not good at reading these threads. Would love to see the numbers included at some point?
